# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [2]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. Also it summarize the number of times the data value occurs.

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [3]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [4]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [5]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [6]:
import pandas as pd
import numpy as np
import scipy.stats as stats

poke = pd.read_csv("pokemon.csv")

poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [7]:
#Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [8]:
#Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts(normalize = True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [9]:
#Create a crosstabs table that shows the frequency of 
#pokemon stage (variable name: Stage) and legendary status (variable name: Legendary).

pd.crosstab(poke["Stage"], poke["Legendary"], normalize = False, margins = False)

Legendary  False  True
Stage                 
1             75     4
2             56     0
3             16     0

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpeg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [10]:
df[["Hours Worked"]]

Hours Worked
0              32
1              43
2              28
3              24
4              25
..            ...
994            40
995            22
996            29
997            22
998            29

[999 rows x 1 columns]

In [11]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [12]:
## Calculate the Median of a column 

df[["Hours Worked"]].median()
# Here 34.0 is even, so it's the avg of the 2 middle values of Hours Worked

Hours Worked    34.0
dtype: float64

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [13]:
## Calculate the Mode of a column 

df[["Hours Worked"]].mode()

Hours Worked
0            38

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [14]:
#Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.
poke["Attack"].mean()

#the avg Attack value is 72.54

72.54966887417218

In [15]:
#Determine the median defense value (Defense). Write a single sentence on what your findings mean.
poke["Defense"].median()

# 65.0 is the middle defense value in our dataset

65.0

In [16]:
#Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean.
poke["Speed"].mode()

#"Speed of value 0 and 1 occur the most in the dataset"

0    70
1    90
dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpeg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpeg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpeg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [17]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [18]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [19]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below 44 hrs, 15% are above this 44 hrs. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [20]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [21]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [22]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [23]:
#Using the pokemon dataset ('poke'), calculate the range for the HP column. 
HP_range = poke["HP"].max() - poke["HP"].min()
print(f"The HP range is {HP_range}")

The HP range is 240


In [24]:
#What is the mean and the standard deviation for the special attack (SpAtk) column?
mean = poke["SpAtk"].mean()
std = poke["SpAtk"].std()

print(f"For the special attack column, the mean is {mean} and the standard deviation is {std}")

For the special attack column, the mean is 67.13907284768212 and the standard deviation is 28.53419930191353


In [25]:
#For the Attack column, determine which value falls at the 80th percentile? 
#Which value falls at the 15th percentile? 
#For each of these values, write a single sentence describing what this means.

poke["Attack"].quantile(.8)
# the value falls at 20% above 

poke["Attack"].quantile(.15)
# Attack of 95 and below of the attack falls into

45.0

In [26]:
#Calculate the second quartile (or the median) for the special defense (SpDef) column. 
#Then, calculate the median of the same column using the method shown above. 
#Are the values the same? 

poke["SpDef"].quantile(.5)
poke["SpDef"].median()

# Both values are the same

65.0

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [27]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [28]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)  # 30 rows by 10 columns

(30, 10)


In [29]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [30]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [31]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [32]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [33]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [34]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [35]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [36]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [37]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [38]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [39]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [40]:
import pandas as pd
import numpy as np
import scipy.stats as stats

bb = pd.read_excel("babies.xlsx")
bb.head()

bwt  gestation parity   age  height  weight      smoke
0  120      284.0     no  27.0    62.0   100.0  nonsmoker
1  113      282.0     no  33.0    64.0   135.0  nonsmoker
2  128      279.0     no  28.0    64.0    77.0     smoker
3  123        NaN     no  36.0    69.0   190.0  nonsmoker
4  108      282.0     no  23.0    67.0   125.0     smoker

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [41]:
bb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   object 
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 67.7+ KB


In [42]:
print(bb.shape)

(1236, 7)


In [43]:
bb_copy = bb.copy
bb.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [44]:
bb.dropna(inplace = True)

In [45]:
bb.isnull().sum()

bwt          0
gestation    0
parity       0
age          0
height       0
weight       0
smoke        0
dtype: int64

In [46]:
print(bb.shape)
bb.head() #new dataset

(1174, 7)


bwt  gestation parity   age  height  weight      smoke
0  120      284.0     no  27.0    62.0   100.0  nonsmoker
1  113      282.0     no  33.0    64.0   135.0  nonsmoker
2  128      279.0     no  28.0    64.0    77.0     smoker
4  108      282.0     no  23.0    67.0   125.0     smoker
5  136      286.0     no  25.0    62.0    93.0  nonsmoker

In [47]:
print("{} rows were dropped from the dataset".format(1236-1174))

62 rows were dropped from the dataset


4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [48]:
#Parity and Smoke because they have string values

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [49]:
bbn = bb.replace(to_replace = ["no", "yes", "nonsmoker", "smoker"], value = [0,1,0,1])
bbn.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0      0
1  113      282.0       0  33.0    64.0   135.0      0
2  128      279.0       0  28.0    64.0    77.0      1
4  108      282.0       0  23.0    67.0   125.0      1
5  136      286.0       0  25.0    62.0    93.0      0

In [50]:
bbnc = bbn.copy()
print(bbn.shape)

(1174, 7)


6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [ ]:
# Outliers
#Outliers are the Dtype int and float

In [52]:
## Calculate quartiles of Outliers

#bwt
q1b = bbn["bwt"].quantile(.25)
q3b = bbn["bwt"].quantile(.75)

print("Q1_bwt:", q1b)
print("Q3_bwt:", q3b)
print(" ")

#gestation
q1g = bbn["gestation"].quantile(.25)
q3g = bbn["gestation"].quantile(.75)

print("Q1_gestation:", q1g)
print("Q3_gestation:", q3g)
print(" ")

#age
q1a = bbn["age"].quantile(.25)
q3a = bbn["age"].quantile(.75)

print("Q1_age:", q1a)
print("Q3_age:", q3a)
print(" ")

#height
q1h = bbn["height"].quantile(.25)
q3h = bbn["height"].quantile(.75)

print("Q1_height:", q1h)
print("Q3_height:", q3h)
print(" ")

#weight
q1w = bbn["weight"].quantile(.25)
q3w = bbn["weight"].quantile(.75)

print("Q1_weight:", q1w)
print("Q3_weight:", q3w)

Q1_bwt: 108.0
Q3_bwt: 131.0
 
Q1_gestation: 272.0
Q3_gestation: 288.0
 
Q1_age: 23.0
Q3_age: 31.0
 
Q1_height: 62.0
Q3_height: 66.0
 
Q1_weight: 114.0
Q3_weight: 139.75


In [53]:
## Calculating the IQR

#bwt
iqrb = q3b - q1b

print("IQR_bwt:", iqrb)
print(" ")

#gestation
iqrg = q3g - q1g

print("IQR_gestation:", iqrg)
print(" ")

#age
iqra = q3a - q1a

print("IQR_age:", iqra)
print(" ")

#height
iqrh = q3h - q1h

print("IQR_height:", iqrh)
print(" ")

#weight
iqrw = q3w - q1w

print("IQR_weight:", iqrw)
print(" ")

IQR_bwt: 23.0
 
IQR_gestation: 16.0
 
IQR_age: 8.0
 
IQR_height: 4.0
 
IQR_weight: 25.75
 


In [54]:
## Determining outlier fences

#bwt
topb = q3b + (iqrb * 1.5)
bottomb = q1b - (iqrb * 1.5)

print("Upper Limit_bwt:", topb)
print("Lower Limit_bwt:", bottomb)
print(" ")

#gestation
topg = q3g + (iqrg * 1.5)
bottomg = q1g - (iqrg * 1.5)

print("Upper Limit_gestation:", topg)
print("Lower Limit_gestation:", bottomg)
print(" ")

#age
topa = q3a + (iqra * 1.5)
bottoma = q1a - (iqra * 1.5)

print("Upper Limit_age:", topa)
print("Lower Limit_age:", bottoma)
print(" ")

#height
toph = q3h + (iqrh * 1.5)
bottomh = q1h - (iqrh * 1.5)

print("Upper Limit_height:", toph)
print("Lower Limit_height:", bottomh)
print(" ")

#weight
topw = q3w + (iqrw * 1.5)
bottomw = q1w - (iqrw * 1.5)

print("Upper Limit_weight:", topw)
print("Lower Limit_weight:", bottomw)
print(" ")

Upper Limit_bwt: 165.5
Lower Limit_bwt: 73.5
 
Upper Limit_gestation: 312.0
Lower Limit_gestation: 248.0
 
Upper Limit_age: 43.0
Lower Limit_age: 11.0
 
Upper Limit_height: 72.0
Lower Limit_height: 56.0
 
Upper Limit_weight: 178.375
Lower Limit_weight: 75.375
 


In [55]:
## Determine the index locations for rows that fall outside of outlier fences

#bwt
iqr_outliersb = bbn.loc[(bb['bwt'] > topb) | (bbn['bwt'] < bottomb)].index

print("INDEX VALUES_bwt:", iqr_outliersb)
print(" ")

#gestation
iqr_outliersg = bbn.loc[(bbn['gestation'] > topg) | (bbn['gestation'] < bottomg)].index

print("INDEX VALUES_gestation:", iqr_outliersg)
print(" ")

#age
iqr_outliersa = bbn.loc[(bb['age'] > topa) | (bbn['age'] < bottoma)].index

print("INDEX VALUES_age:", iqr_outliersa)
print(" ")

#height
iqr_outliersh = bbn.loc[(bbn['height'] > toph) | (bbn['height'] < bottomh)].index

print("INDEX VALUES_height:", iqr_outliersh)
print(" ")

#weight
iqr_outliersw = bb.loc[(bbn['weight'] > topw) | (bbn['weight'] < bottomw)].index

print("INDEX VALUES_weight:", iqr_outliersw)
print(" ")

INDEX VALUES_bwt: Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169],
           dtype='int64')
 
INDEX VALUES_gestation: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')
 
INDEX VALUES_age: Int64Index([633, 1070], dtype='int64')
 
INDEX VALUES_height: Int64Index([434, 1208], dtype='int64')
 
INDEX VALUES_weight: Int64Index([  14,   23,   88,  117,  125,  149,  162,  181,  183,  2

In [56]:
bbn.iloc[[ 14,   23,   88,  117,  125,  149,  162,  181,  183,  222,  240,
             287,  314,  344,  411,  426,  458,  512,  522,  528,  563,  595,
             608,  622,  632,  723,  733,  747,  849,  858,  859,  865,  888,
             911,  924, 1008, 1021, 1148, 1167]]

bwt  gestation  parity   age  height  weight  smoke
15    115      285.0       0  38.0    63.0   130.0      0
24    131      294.0       0  23.0    65.0   122.0      0
94     85      278.0       0  23.0    61.0   103.0      1
127    94      267.0       0  30.0    62.0   120.0      1
135   117      258.0       0  31.0    64.0   120.0      0
162   125      262.0       0  36.0    66.0   190.0      0
176   134      267.0       0  30.0    66.0   170.0      0
197   118      276.0       0  22.0    66.0   147.0      1
199   137      291.0       0  34.0    61.0   110.0      0
241   114      278.0       0  25.0    65.0   140.0      1
262   124      287.0       0  27.0    62.0   105.0      1
309    71      281.0       0  32.0    60.0   117.0      1
338   111      270.0       0  22.0    59.0   103.0      0
371   124      294.0       0  26.0    62.0   122.0      0
442   100      275.0       0  27.0    64.0   111.0      1
458    85      258.0       0  41.0    67.0   399.0      0
493   115      290.0       1  19.0    65.0   118.0      0
549   122      292.0       1  25.0    65.0   125.0      0
559   124      285.0       1  27.0    63.0   114.0      0
565   153      297.0       0  27.0    66.0   145.0      0
601    85      270.0       1  19.0    63.0   118.0      1
633   143      294.0       0  44.0    65.0   145.0      0
647    93      270.0       0  25.0    64.0   125.0      1
664   127      288.0       1  20.0    65.0   115.0      1
676   100      275.0       0  26.0    60.0   115.0      0
772   112      277.0       1  23.0    64.0   118.0      0
782   148      286.0       0  38.0    68.0   160.0      0
796   152      286.0       1  19.0    67.0   135.0      0
902   112      292.0       1  28.0    62.0   110.0      1
911   138      280.0       1  30.0    65.0   478.0      0
912    58      245.0       0  34.0    64.0   156.0      1
918   142      284.0       1  31.0    66.0   137.0      1
942    89      275.0       0  34.0    66.0   170.0      0
966   120      283.0       0  28.0    64.0   122.0      1
980   123      283.0       0  21.0    65.0   110.0      0
1066  118      279.0       1  21.0    64.0   108.0      0
1079  119      286.0       1  33.0    67.0   137.0      0
1209  144      283.0       1  25.0    66.0   140.0      0
1229  127      290.0       0  27.0    65.0   121.0      0

7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [57]:
#Frequencies
bb["parity"].value_counts()

no     866
yes    308
Name: parity, dtype: int64

In [58]:
bb["smoke"].value_counts()

nonsmoker    715
smoker       459
Name: smoke, dtype: int64

In [59]:
## Relative frequencies
bb["parity"].value_counts(normalize = True)

no     0.737649
yes    0.262351
Name: parity, dtype: float64

In [60]:
bb["smoke"].value_counts(normalize = True)

nonsmoker    0.609029
smoker       0.390971
Name: smoke, dtype: float64

In [61]:
## a crosstab table using both the categorical variables 
pd.crosstab(bb["parity"], bb["smoke"], normalize = False, margins = False)

smoke   nonsmoker  smoker
parity                   
no            525     341
yes           190     118

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [62]:
## Calculating the Measures of Central Tendency

#mean
meanb = bb["bwt"].mean()
meang = bb["gestation"].mean()
meana = bb["age"].mean()
meanh = bb["height"].mean()
meanw = bb["weight"].mean()

print("The mean of 'bwt' is :", meanb)
print("The mean of 'gestation' is :", meang)
print("The mean of 'age' is :", meana)
print("The mean of 'height' is :", meanh)
print("The mean of 'weight' is :", meanw)
print(" ")

#mode
modeb = bb['bwt'].mode()
modeg = bb['gestation'].mode()
modea = bb['age'].mode()
modeh = bb['height'].mode()
modew = bb['weight'].mode()

print("The mode of 'bwt' is :", modeb)
print("The mode of 'gestation' is :", modeg)
print("The mode of 'age' is :", modea)
print("The mode of 'height' is :", modeh)
print("The mode of 'weight' is :", modew)
print(" ")

#median
medianb = bb['bwt'].median()
mediang = bb['gestation'].median()
mediana = bb['age'].median()
medianh = bb['height'].median()
medianw = bb['weight'].median()

print("The median of 'bwt' is :", medianb)
print("The median of 'gestation' is :", medianb)
print("The median of 'age' is :", medianb)
print("The median of 'height' is :", medianb)
print("The median of 'weight' is :", medianb)

The mean of 'bwt' is : 119.46252129471891
The mean of 'gestation' is : 279.1013628620102
The mean of 'age' is : 27.228279386712096
The mean of 'height' is : 64.04940374787053
The mean of 'weight' is : 131.32623509369677
 
The mode of 'bwt' is : 0    115
dtype: int64
The mode of 'gestation' is : 0    282.0
dtype: float64
The mode of 'age' is : 0    23.0
dtype: float64
The mode of 'height' is : 0    64.0
dtype: float64
The mode of 'weight' is : 0    130.0
dtype: float64
 
The median of 'bwt' is : 120.0
The median of 'gestation' is : 120.0
The median of 'age' is : 120.0
The median of 'height' is : 120.0
The median of 'weight' is : 120.0


In [63]:
# The avg birth weight of newborn baby is 119.5 lbs
# The avg gestation length is 279 weeks
# The avg age of mother is 27
# The avg height of mother is 64 inches
# The avg weight of mother is 131 lbs

## Respectively, the values that occurs the most in columns "bwt", "gestation", "age", "height" and "weight"
## are "blanks" (all across), and 115, 282, 23, 64, and 130

### 120 is the center value in all 5 categorical variables' set.

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [64]:
# Range
bwt_range = bb['bwt'].max() - bb['bwt'].min()
gestation_range = bb['gestation'].max() - bb['gestation'].min()
age_range = bb['age'].max() - bb['age'].min()
height_range = bb['height'].max() - bb['height'].min()
weight_range = bb['weight'].max() - bb['weight'].min()

print("\t\tThe Range")
print(f"The range of weight in 'birth weight of newborn baby': {bb['bwt'].max()} - {bb['bwt'].min()} = {bwt_range}")
print(f"The range of weeks in 'gestation length': {bb['gestation'].max()} - {bb['gestation'].min()} = {gestation_range}")
print(f"The range of age in 'age of mother': {bb['age'].max()} - {bb['age'].min()} = {age_range}")
print(f"The range of height in 'height of mother':{bb['height'].max()} - {bb['height'].min()} = {height_range}" )
print(f"The range of weight in 'weight of mother': {bb['weight'].max()} - {bb['weight'].min()} = {weight_range}")
print(" ")

# Standard Deviation
stdb = bb["bwt"].std()
stdg = bb["gestation"].std()
stda = bb["age"].std()
stdh = bb["height"].std()
stdw = bb["weight"].std()

print("\t\tThe Standard deviation")
print("The standard deviation 'bwt':", stdb)
print("The standard deviation 'gestation':", stdg)
print("The standard deviation 'age ':", stda)
print("The standard deviation 'height':", stdh)
print("The standard deviation 'weight':", stdw)
print(" ")

# 85th Percentile
percb = np.percentile(bb['bwt'], 85)
percg = np.percentile(bb['gestation'], 85)
perca = np.percentile(bb['age'], 85)
perch = np.percentile(bb['height'], 85)
percw = np.percentile(bb['weight'], 85)

print("\t\tThe 85th Percentile")
print(f"""{percb} falls at the 85th percentile of 'bwt'
      which means that 85% of the values in 'bwt 'are below {percb} lbs , 15% are above {percb} lbs""")
print(f"""{percg} falls at the 85th percentile of 'gestation'
      which means that 85% of the values in 'gestation' are below {percg} lbs , 15% are above {percg} lbs""")
print(f"""{perca} falls at the 85th percentile of 'age'
      which means that 85% of the values in 'age'are below {perca} years of age , 15% are above {perca} years old""")
print(f"""{perch} falls at the 85th percentile of 'height'
      which means that 85% of the values in 'height'are below {perch} inches , 15% are above {perch} inches""")
print(f"""{percw} falls at the 85th percentile of 'weight'
      which means that 85% of the values in 'weight'are below {percw} lbs , 15% are above {percw} lbs""")

		The Range
The range of weight in 'birth weight of newborn baby': 176 - 55 = 121
The range of weeks in 'gestation length': 353.0 - 148.0 = 205.0
The range of age in 'age of mother': 45.0 - 15.0 = 30.0
The range of height in 'height of mother':72.0 - 53.0 = 19.0
The range of weight in 'weight of mother': 1500.0 - 27.0 = 1473.0
 
		The Standard deviation
The standard deviation 'bwt': 18.32867144332098
The standard deviation 'gestation': 16.010305127129392
The standard deviation 'age ': 5.817838716855792
The standard deviation 'height': 2.5261015226020365
The standard deviation 'weight': 54.1038284756229
 
		The 85th Percentile
138.0 falls at the 85th percentile of 'bwt'
      which means that 85% of the values in 'bwt 'are below 138.0 lbs , 15% are above 138.0 lbs
292.0 falls at the 85th percentile of 'gestation'
      which means that 85% of the values in 'gestation' are below 292.0 lbs , 15% are above 292.0 lbs
34.0 falls at the 85th percentile of 'age'
      which means that 85% of t

In [65]:
###Which column has the largest range? Which has the smallest? What does this mean?
# The "weight of mother" column has the largest range, whereas "age of mother" has the smallest one.
# it means that there is a large diff between the highest and lowest values in column "weight of mother"
# and the opposite is the truth for column "age of mother"

###Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

# column "weight of mother" has the largest std, which indicates that the data points in the set is spread out over a wider range of values.
# The opposite is noticed in column "height of mother" where the data points are clustered around the mean.